flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java

```diff
@@ -173,72 +174,192 @@ private static void assertResponse(
+                        } catch (IOException e) {
+                            throw new CompletionException(e);
+                        }
+                    });
+        }
+
+        /** Closes the established connection from the server side. */
+        public void closeAcceptingSocket() throws Exception {
+            if (connectionFuture.isDone()) {
+                connectionFuture.get().close();
+                connectionFuture = new CompletableFuture<>();
             }
         }
 
-        public void close() {
-            running = false;
+        /**
+         * Waits for the connection to be established between the client and the server. This method
+         * will block until a request is sent by the client and a {@code handle*} call is initiated
+         * by this instance.
+         */
+        public void waitForConnectionToBeEstablished()
+                throws ExecutionException, InterruptedException {
+            connectionFuture.get();
+        }
+
+        /**
+         * Handles a request with the given data in a synchronous fashion. The {@code
+         * TestingSocketServer}'s default meta information is attached to the response.
+         *
+         * @see #DEFAULT_SERVER_RESPONSE_VERSION
+         * @see #DEFAULT_SERVER_RESPONSE_OFFSET
+         */
+        public void handleRequest(List<Row> actualData) {
+            handleRequest(
+                    DEFAULT_SERVER_RESPONSE_VERSION, DEFAULT_SERVER_RESPONSE_OFFSET, actualData);
+        }
+
+        /**
+         * Handles the next request synchronously. The passed {@code actualData} will be forwarded
+         * to the response.
+         */
+        public void handleRequest(String actualVersion, int actualOffset, List<Row> actualData) {
+            handleRequestAsync(
+                            actualVersion,
+                            actualOffset,
+                            CompletableFuture.completedFuture(actualData))
+                    .join();
+        }
+
+        /**
+         * Handles a request with the given data in an asynchronous fashion with the {@code
+         * TestingSocketServer}'s default meta information being attached to the response. The
+         * returned future completes when the server side's processing completes.
+         *
+         * @see #DEFAULT_SERVER_RESPONSE_VERSION
+         * @see #DEFAULT_SERVER_RESPONSE_OFFSET
+         */
+        public CompletableFuture<Void> handleRequestAsync(
+                CompletableFuture<List<Row>> actualDataAsync) {
+            return handleRequestAsync(
+                    DEFAULT_SERVER_RESPONSE_VERSION,
+                    DEFAULT_SERVER_RESPONSE_OFFSET,
+                    actualDataAsync);
+        }
+
+        private CompletableFuture<SocketConnection> connectSocket() {
+            if (!connectionFuture.isDone()) {
+                FutureUtils.forward(acceptSocketAsync(), connectionFuture);
```

Comment:
Feels weird to set up this forwarding here instead of whenever the `connectionFuture` is reset (in the constructor and closeAcceptingSocket)

Reply:
That remark, I don't get. We want to have the same future that's accessible all the time (that's why `new CompletableFuture` here and in the constructor) but we do want to trigger the socket listening later on. How would you implement that differently. It feels natural to me. But I might have missed your point.

Reply:
I guess my point is why accepting the connection is a manual step in the first place.

This can also leak connections if connectSocket is called multiple times before the connectionFuture is being completed.